src/H5FDsubfiling/H5FDsubfiling.c

```diff
@@ -2581,138 +2043,913 @@ H5FD__subfiling_ctl(H5FD_t *_file, uint64_t op_code, uint64_t flags, const void
+     * of subfiles.
+     *
+     * Given the current I/O and the I/O concentrator info,
+     * we can determine some I/O transaction parameters.
+     * In particular, for large I/O operations, each IOC
+     * may require multiple I/Os to fulfill the user I/O
+     * request. The block size and number of IOCs are used
+     * to size the vectors that will be used to invoke the
+     * underlying I/O operations.
+     */
+    if (NULL == (sf_context = (subfiling_context_t *)H5_get_subfiling_object(file_ptr->context_id)))
+        H5_SUBFILING_GOTO_ERROR(H5E_INTERNAL, H5E_BADVALUE, FAIL,
+                                "invalid or missing subfiling context object");
+    assert(sf_context->topology);
+
+    if ((num_subfiles = sf_context->sf_num_subfiles) <= 0)
+        H5_SUBFILING_GOTO_ERROR(H5E_INTERNAL, H5E_BADVALUE, FAIL, "invalid number of subfiles (%d)",
+                                num_subfiles);
+
+    if (num_subfiles == 1) {
+        uint32_t u32_io_count;
+
+        /***************************************
+         * No striping - just a single subfile *
+         ***************************************/
+
+        /*
+         * Convert the I/O count back to a uint32_t for the vector I/O
+         * call until the interface can possibly be changed to use size_t
+         * in the future
+         */
+        H5_CHECKED_ASSIGN(u32_io_count, uint32_t, io_count, size_t);
+
+        if (io_type == IO_TYPE_WRITE) {
+            /* Make vector write call to VFD controlling subfiles */
+            if (H5FD_write_vector(file_ptr->sf_file, u32_io_count, types, addrs, sizes, (const void **)bufs) <
+                0)
+                H5_SUBFILING_GOTO_ERROR(H5E_IO, H5E_WRITEERROR, FAIL, "write to subfile failed");
+
+            /*
+             * Mirror superblock writes to the stub file so that
+             * legacy HDF5 applications can check what type of
+             * file they are reading
+             */
+            if (H5FD__subfiling_mirror_writes_to_stub(file_ptr, u32_io_count, types, addrs, sizes,
+                                                      (const void **)bufs) < 0)
+                H5_SUBFILING_GOTO_ERROR(H5E_IO, H5E_WRITEERROR, FAIL, "mirrored write to stub file failed");
+        }
+        else {
+            /* Make vector read call to VFD controlling subfiles */
+            if (H5FD_read_vector(file_ptr->sf_file, u32_io_count, types, addrs, sizes, (void **)bufs) < 0)
+                H5_SUBFILING_GOTO_ERROR(H5E_IO, H5E_READERROR, FAIL, "read from subfile failed");
+        }
+    }
+    else {
+        uint32_t iovec_len;
+        size_t   ioreq_count = 0;
+        herr_t   status;
+
+        /*************************************
+         * Striping across multiple subfiles *
+         *************************************/
+
+        /*
+         * Generate the types, addrs, sizes and bufs I/O vectors for
+         * this I/O request.
+         */
+        status = generate_io_vectors(
```

Comment:
The old `init_indep_io` function has been replaced by `generate_io_vectors`, which can now handle translating more than one (offset, I/O size, type, buffer) tuple into a set of I/O vectors that spans the subfiles.